tools/sesman/sesmanworker/engine.py

```diff
@@ -1760,3 +1774,198 @@ def get_target_str(self):
+    
+    We can combine 2 filters or more like above but we 
+    can also execute only one filter on target.
+    The following target filter strings are accepted :
+    "$account:my_account"
+    "$domain:my_domain"
+    "$device:my_device"
+    "$service:my_service"
+    "$auth:my_author"
+
+    There is no specifc order for apply filters. 
+    The following target filter string is also valid :
+    "$device:my_device,$account:my_account,$domain:my_domain"
+
+    However, target filter string badly formatted like containing 
+    duplicated filter keywords, missing values and others aren't 
+    allowed and will be skipped and considered as no valid.
+    The following examples are forbidden :
+    "$account:my_account,$account:my_account"
+    "$domain:my_domain,"
+    "$my_device:my_device"
+    "$service:"
+    "$abcd:my_author"
+
+    """
+    
+    __slots__ = ("filter_pattern_dict")
+    
+    def __init__(self) :
+        self.filter_pattern_dict = {
+            "account" : None,
+            "domain" : None,
+            "device" : None,
+            "service" : None,
+            "auth" : None
+        }
+
+    def parse_target_filter_pattern(self, target_filter_pattern) :
+        FILTER_KV_SEPARATOR = ':'
+        FILTER_SEPARATOR = ','
+        filter_kv_list = target_filter_pattern.split(FILTER_SEPARATOR)
+
+        for filter_kv in filter_kv_list :
+            token_list = filter_kv.split(FILTER_KV_SEPARATOR, 1)
+
+            if len(token_list) != 2 :
+                raise ParsingError(filter_kv)
+
+            filter_keyword, filter_value = tuple(token_list)
+
+            if not filter_value :
+                raise ParsingError(filter_kv)
+
+            is_filter_keyword, filter_keyword_no_prefix = (
+                self._is_filter_keyword(filter_keyword))
+
+            if not is_filter_keyword :
+                raise FilterKeywordSyntaxError(filter_keyword)
+            if self.filter_pattern_dict[filter_keyword_no_prefix] is None :
+                self.filter_pattern_dict.update(
+                    {filter_keyword_no_prefix : filter_value})
+            else :
+                raise FilterValueDuplicateError(filter_keyword)
+
+    def is_filterable(self, target) :
+        if not self.filter_pattern_dict.values() :
+            raise NoFilterValueError(target)
+
+        target_field_dict = {}
+
+        target_field_dict.fromkeys(range(len(self.filter_pattern_dict)))
+        self._get_fields_from_target(target, target_field_dict)
+        for filter_keyword, filter_value in self.filter_pattern_dict.items() :
+            if filter_value is None :
+                continue
+            if not self._filter_pattern_is_applicable(filter_keyword,
+                                                      filter_value,
+                                                      target_field_dict) :
+                return False
+        return True
```

Comment:
```py
return all(self._filter_pattern_is_applicable(filter_keyword, filter_value, target_field_dict)
           for filter_keyword, filter_value in self.filter_pattern_dict.items()
           if filter_value is not None)
```
Can `filter_value` really be `None` ?

Reply:
Yes : 

It's for the following case if user type filter string like "$device:my_device" and on target list, we have a target like my_account@my_device:service. If I don't check filter_value is None with "$service" filter_keyword (filter_value is set to None at TargetAccurateFilter constructor which means no filter for filter_keyword) and don't execute "continue", the loop will stop by returning "False" as filter for "$service" doesn't exist

Reply:
However, `_get_fields_from_target` never puts `None` in the dictionary and as line 1893 does nothing, the dictionary is empty before this call.
Also, this function should probably return a dictionary/iterator rather than take an output parameter.